README.md
# Objective-Clean

Objective-Clean is a VS Code extension to modernize old Objective-C code. We've all been there: you crack open a project written by someone who learned Objc 20 years ago and never got into the habit of using collection literals and subscripting syntax, and you wish you could flip a switch to convert that code to use modern Objective-C syntax!

Well, now you can!

## Features

This extension comes with a single command, `Convert File to Modern Objective-C Syntax`, that will only do the following:

- Convert `NSNumber` initializers to use `NSNumber` literals
- Convert `NS[Mutable]Dictionary|Array` initializers to use collection literals
- Convert `NS[Mutable]Dictionary|Array` accessors to use subscripting syntax

https://user-images.githubusercontent.com/8371943/119164357-f5798100-ba21-11eb-9615-1f94564b4fcf.mp4

## Known Issues

For most operations, this extension uses simple regex find and replace. As a result, there may be edge cases it cannot handle or won't handle correctly without a full-blown parser. Some contrived examples:

- `[[foo bar] objectForKey:key]` may not be handled, in order for `[object method:[dict objectForKey:key]]` to work
- `[[foo bar:foo[1]] objectForKey:key]` may result in `[[foo bar:foo1]][key]`

## TODO

- [ ] Implement stricter parsing for dictionary initializers
- [ ] Implement a parser for collection accessors instead of using regex
- [x] Split all functionality into individual commands
  - [x] Add a command to run a list of other commands of this extension
  - [x] Add setting to choose "default" commands to run in the above setting

Add new commands for the following:
- [x] alloc-init calls to `new`
- [x] Collection `initWithCapacity:` calls to `new`/alloc-init
- [ ] Accesses of known class properties to dot syntax (i.e. `UIColor.redColor`, `UIDevice.currentDevice`)
- [x] `[[UIView alloc] initWithFrame:CGRectZero]` to `new`/alloc-init
- [x] `CGRect(0, 0, 0, 0)` to `CGRectZero`
- [ ] Remove unnecessary casts in message sends, such as `[(NSNumber *)foo boolValue]`
- [ ] Enforce proper method declaration format, i.e. `-(void) foo: (id ) bar` to `- (void)foo:(id)bar`

## Release Notes

See the [changelog](CHANGELOG.md)


CHANGELOG.md
## Initial release | 2021-05-20
- Adds a single command to do the following:
  - Convert `NSNumber` initializers to use `NSNumber` literals
  - Convert `NS[Mutable]Dictionary|Array` initializers to use collection literals
  - Convert `NS[Mutable]Dictionary|Array` accessors to use subscripting syntax

## 2.0.0 | 2021-05-22
- Implement many new commands
- Add settings to enable or disable certain rules
- Add ability to preview changes in a diff first
- Add ability to run all rules or only enabled rules

## 2.1.0 | 2021-08-07
- Support converting C-style block comments to triple-slash comments
- Fix a bug where no-preview mode wouldn't change the file correctly


src/test/suite/extension.test.ts
//
// Note: This example test is leveraging the Mocha test framework.
// Please refer to their documentation on https://mochajs.org/ for help.
//

import * as assert from 'assert';
import Modernizer from '../../modernizer';

const testString = `
- (void)method {
    NSMutableDictionary *dict = [NSMutableDictionary dictionaryWithObjectsAndKeys:
        self.filterName, @"filterName",
        self.filterText, @"filterText",
        [NSNumber numberWithInt:self.blockType], @"blockType",
        [NSNumber numberWithInt:self.filterType], @"filterType",
        [NSNumber numberWithBool:self.onSchedule], @"onSchedule",
        [NSNumber numberWithBool:self.whitelistMode], @"whitelistMode",
        [NSNumber numberWithBool:self.showInNotificationCenter], @"showInNotificationCenter",
        self.startTime, @"startTime",
        self.endTime, @"endTime",
        self.weekDays, @"weekDays",
    nil];
  
    NSArray *planets = [NSArray arrayWithObject: @"earth"];
    NSArray *months = [NSArray arrayWithObjects:
        @"Jan", @"Feb",
        @"Mar", @"Apr,
        @"May", @"Jun",
        @"Jul", @"Aug",
        @"Sept, @"Oct",
        @"Nov", @"Dec",
        nil];
    NSArray *weekdays = [NSArray arrayWithObjects: @"Sun", @"Mon", @"Tue", @"Wed", @"Thu",@"Fri", @"Sat", nil];
    id monday = [weekdays objectAtIndex:1];
    
    NSMutableArray *mutableWeekdays = weekdays.mutableCopy;
    [mutableWeekdays replaceObjectAtIndex:3 withObject:monday];
    
    self.filterName = [dict objectForKey:@"filterName"];
    self.filterType = [((NSNumber*)[dict objectForKey:@"filterType"]) intValue];

    if ([dict objectForKey:@"appToBlockIdentifier"] != nil) {
        AppInfo *app = [[AppInfo alloc] init];
        app.appIdentifier = [dict objectForKey:@"appToBlockIdentifier"];
        app.appName = [dict objectForKey:@"appToBlockName"];
        self.appToBlock = app;
    }
    
    self.tableView = [[UITableView alloc] initWithFrame:CGRectMake(0,0,screenWidth,screenHeight) style:UITableViewStyleGrouped];
    self.tableView = [[UITableView alloc] initWithFrame:CGRectMake(0,0,screenWidth,screenHeight)];
    self.tableView = [[UITableView alloc] initWithFrame:CGRectMake(0,0, 0.0, 0)];
    
    self.onSchedule = [((NSNumber*)[dict objectForKey:@"onSchedule"]) boolValue];
    self.whitelistMode = [((NSNumber*)[dict objectForKey:@"whitelistMode"]) boolValue];
    self.showInNotificationCenter = [((NSNumber*)[dict objectForKey:@"showInNotificationCenter"]) boolValue];

    self.startTime = [dict objectForKey:@"startTime"];
    self.endTime = [dict objectForKey:@"endTime"];
    self.weekDays = [NSMutableArray arrayWithArray:[dict objectForKey:@"weekDays"]];
    
    id five = [NSNumber numberWithInteger:5];
    id pi = [NSNumber numberWithDouble:3.14];
    id yes = [NSNumber numberWithBool:YES];
    id no = [NSNumber numberWithBool:NO];
    id filtertype = [NSNumber numberWithInt:self.filterType];
    id view = [[UIButton alloc] initWithFrame:CGRectMake(0, 0.0, 0, 0)];
    CGRect *rects = {
        CGRectMake(0,0,0,0),
        CGRectMake(0, 0, 0, 0),
        CGRectMake(0.0,0.0,0.0,0.0),
        CGRectMake(0.0, 0.0, 0.0, 0.0),
        CGRectMake(0.0 , 0, 0.f, 0.0f),
    };
    
    id five = [NSNumber numberWithInteger:5];
    id pi = [NSNumber numberWithDouble:3.14];
    id yes = [NSNumber numberWithBool:YES];
    id no = [NSNumber numberWithBool:NO];
    id filtertype = [NSNumber numberWithInt:self.filterType];
    id view = [[UIButton alloc] initWithFrame:CGRectMake(0, 0.0, 0, 0)];
    CGRect *rects = {
        CGRectMake(0,0,0,0),
        CGRectMake(0, 0, 0, 0),
        CGRectMake(0.0,0.0,0.0,0.0),
        CGRectMake(0.0, 0.0, 0.0, 0.0),
        CGRectMake(0.0 , 0, 0.f, 0.0f),
    };
}
`;

const testStringFixed = `
- (void)method {
    NSMutableDictionary *dict = @{
        @"filterName": self.filterName,
        @"filterText": self.filterText,
        @"blockType": @(self.blockType),
        @"filterType": @(self.filterType),
        @"onSchedule": @(self.onSchedule),
        @"whitelistMode": @(self.whitelistMode),
        @"showInNotificationCenter": @(self.showInNotificationCenter),
        @"startTime": self.startTime,
        @"endTime": self.endTime,
        @"weekDays": self.weekDays,
    }.mutableCopy;
  
    NSArray *planets = @[@"earth"];
    NSArray *months = @[
        @"Jan", @"Feb",
        @"Mar", @"Apr,
        @"May", @"Jun",
        @"Jul", @"Aug",
        @"Sept, @"Oct",
        @"Nov", @"Dec",
        ];
    NSArray *weekdays = @[@"Sun", @"Mon", @"Tue", @"Wed", @"Thu",@"Fri", @"Sat", ];
    id monday = weekdays[1];
    
    NSMutableArray *mutableWeekdays = weekdays.mutableCopy;
    mutableWeekdays[3] = monday;
    
    self.filterName = dict[@"filterName"];
    self.filterType = [((NSNumber*)dict[@"filterType"]) intValue];

    if (dict[@"appToBlockIdentifier"] != nil) {
        AppInfo *app = [AppInfo new];
        app.appIdentifier = dict[@"appToBlockIdentifier"];
        app.appName = dict[@"appToBlockName"];
        self.appToBlock = app;
    }
    
    self.tableView = [[UITableView alloc] initWithFrame:CGRectMake(0,0,screenWidth,screenHeight) style:UITableViewStyleGrouped];
    self.tableView = [[UITableView alloc] initWithFrame:CGRectMake(0,0,screenWidth,screenHeight)];
    self.tableView = [UITableView new];
    
    self.onSchedule = [((NSNumber*)dict[@"onSchedule"]) boolValue];
    self.whitelistMode = [((NSNumber*)dict[@"whitelistMode"]) boolValue];
    self.showInNotificationCenter = [((NSNumber*)dict[@"showInNotificationCenter"]) boolValue];

    self.startTime = dict[@"startTime"];
    self.endTime = dict[@"endTime"];
    self.weekDays = [NSMutableArray arrayWithArray:dict[@"weekDays"]];
    
    id five = @5;
    id pi = @3.14;
    id yes = @YES;
    id no = @NO;
    id filtertype = @(self.filterType);
    id view = [UIButton new];
    CGRect *rects = {
        CGRectZero,
        CGRectZero,
        CGRectZero,
        CGRectZero,
        CGRectZero,
    };
    
    id five = @5;
    id pi = @3.14;
    id yes = @YES;
    id no = @NO;
    id filtertype = @(self.filterType);
    id view = [UIButton new];
    CGRect *rects = {
        CGRectZero,
        CGRectZero,
        CGRectZero,
        CGRectZero,
        CGRectZero,
    };
}
`;

const mangledCasesOriginal = `
[NSDictionary dictionaryWithObjectsAndKeys:[foo bar:@"]"], @"key", nil];
[[foo bar:foo[1]] objectForKey:key];
`;

const mangledCasesMangled = `
@{[foo bar:@"]"], @"key", };
[[foo bar:foo1]][key];
`;

const testComments = `
if (foo) {
    /*
     * Block comment
     * 
     * blah blah
     */
    return bar;
} else {
    /** inline comment */
    return baz;
    
    /*
    Block comment with no
    leading space or *
    */
}

/** Weirdly formatted comment
 * second line
 * third line
 * 
 * blank line above
 */
void main() { }
`;
const testCommentsFixed = `
if (foo) {
    // Block comment
    // 
    // blah blah
    return bar;
} else {
    /// inline comment 
    return baz;
    
    // Block comment with no
    // leading space or *
}

/// Weirdly formatted comment
/// second line
/// third line
/// 
/// blank line above
void main() { }
`;

// const unhandledCases = `
// `;

suite("Extension Tests", () => {

    test("Modernize test strings", () => {
        const cases = {
            [testString]: testStringFixed,
            [mangledCasesOriginal]: mangledCasesMangled,
            [testComments]: testCommentsFixed,
            // [unhandledCases]: unhandledCases
        };
        
        for (const [test, expectation] of Object.entries(cases)) {
            const cleaned = Modernizer.modernize(test);
            assert.strictEqual(cleaned, expectation);
        }
    });
});


src/scanner.ts
//
//  scanner.ts
//  Objective-Clean
//  
//  Created by Tanner Bennett on 2021-05-20
//  Copyright © 2021 Tanner Bennett. All rights reserved.
//

// import { assert } from "console";

function assert(condition: boolean): asserts condition {
    if (!condition) {
        throw new Error("Assertion failed");
    }
}

type Tags = { [key: string]: string | undefined };

export class Scanner {
    private input: string;
    public location = 0;
    
    constructor(input: string) {
        this.input = input;
    }
    
    get string(): string {
        return this.input.slice(this.location);
    }
    
    get scanned(): string {
        return this.scannedSince(0);
    }
    
    get remaining(): string {
        return this.input.substr(this.location);
    }
    
    get isAtEnd(): boolean {
        return this.location >= this.input.length;
    }
    
    undo(loc: number): false {
        this.location = loc;
        return false;
    }
    
    scannedSince(loc: number): string {
        return this.input.slice(loc, this.location);
    }
    
    scanString(str: string): boolean {
        if (this.string.startsWith(str)) {
            this.location += str.length;
            return true;
        }
        
        return false;
    }
    
    scanUpToString(str: string): string | false {
        const idx = this.string.indexOf(str);
        if (idx != -1) {
            const scanned = this.string.slice(0, idx);
            this.location += idx;
            return scanned;
        }
        
        return false;
    }
    
    scanUpToPattern(pattern: RegExp): string | false {
        const idx = this.string.search(pattern);
        if (idx != -1) {
            const scanned = this.string.slice(0, idx);
            this.location += idx;
            return scanned;
        }
        
        return false;
    }
    
    scanPattern(pattern: RegExp): string | false {
        const match = this.string.match(pattern);
        if (!match || match.index != 0) {
            return false;
        }
        
        this.location += match[0].length;
        return match[0];
    }
    
    scanWhitespace(): Scanner {
        this.scanPattern(/\s*/);
        return this;
    }
    
    scanPastTagPair(opening: string, closing: string): boolean {
        // const openings = new RegExp(`(${Object.keys(possibleTags).join('|')})`);
        
        // Sanity check: we start with the opening tag
        if (!this.scanString(opening)) {
            return false;
        }
        
        // Look for the closing tag; note that this will not do advanced
        // parsing and could mess up on something like [foo bar:@"]"]
        let remaining = 1, i = this.location;
        for (i = this.location; remaining > 0 && i < this.input.length; i++) {
            const ch = this.input[i];
            // Check closing first in case opening == closing
            if (ch == closing) {
                remaining--;
            } else if (ch == opening) {
                remaining++;
            }
        }
        
        if (remaining > 0) {
            return false;
        }
        
        this.location = i;
        return true;
    }
    
    scanUpToStringAfterTags(str: string, trimLeadingWhitespace: boolean = false): string | false {
        // For aborting entirely
        const _backup = this.location;
        // For returning the scanned string
        const start = trimLeadingWhitespace ? this.scanWhitespace().location : this.location;
        
        // Sanity check: do we even contain the search string?
        if (!this.string.includes(str)) {
            return false;
        }
        
        const tags: Tags = {
            '[': ']',
            '{': '}',
            '(': ')',
            '<': '>',
            "'": "'",
            '"': '"',
            '/*': '*/'
        };
        
        const tagsOrSearchString = new RegExp(`([\[\{\(<'"]|\/\\*|${str})`);
        const openings = /([\[\{\(<'"]|\/\*)/;
        
        // Repeatedly scan up-to the braces or search string until one
        // is found. The loop terminates when the search string is found.
        while (!this.string.startsWith(str)) {
            const currString = this.string;
            
            // Scan up to either
            this.scanUpToPattern(tagsOrSearchString);
            
            // Did we find a tag?
            const opening = this.scanPattern(openings);
            if (opening !== false) {
                const closing = tags[opening];
                assert(closing != undefined);
                
                // Scan past it (back-up over the opening tag first)
                this.location -= opening.length;
                if (!this.scanPastTagPair(opening, closing)) {
                    // Abort if we don't scan a complete tag
                    return this.undo(_backup);
                }
            }
            
            // Loop will exit if opening is not found, which
            // means we finally arrived at the search string.
        }
        
        // Return thie scanned string
        return this.scannedSince(start);
    }
}


src/extension.ts
//
//  extension.ts
//  Objective-Clean
//  
//  Created by Tanner Bennett on 2021-05-19
//  Copyright © 2021 Tanner Bennett. All rights reserved.
//

'use strict';
import * as vscode from 'vscode';
import Modernizer, { OCMCommand, OCMCompoundRuleCommand } from './modernizer';

function wantsDiffPreview(): boolean {
    return vscode.workspace.getConfiguration('objective-clean').get('behavior') == 'preview';
}

async function processDocument(cmd: OCMCommand) {
    const editor = vscode.window.activeTextEditor;
    if (editor) {
        const document = editor.document;
        
        if (wantsDiffPreview()) {
            // Generate an objc-clean scheme'd URI for the modified file
            const orig = document.uri;
            const diff = orig.with({ scheme: 'obj-clean', query: `?cmd=${cmd}` });
            const filename = orig.path.split('/').pop();
            const title = `Obj-Clean: ${filename}`;
            // Present a diff of the active file and the modified file
            vscode.commands.executeCommand('vscode.diff', document.uri, diff, title);
        } else {
            // Modify the active file immediately
            const changes = Modernizer.applyCommandToText(cmd, editor.document.getText());
            Modernizer.shared.applyChangesToTextEditor(changes, editor);
        }
    }
}

export function activate(context: vscode.ExtensionContext) {
    // Register document content provider for diffing
    context.subscriptions.push(
        vscode.workspace.registerTextDocumentContentProvider(
            'obj-clean', Modernizer.shared
        )
    );
    
    // Individual rules (includes the 'apply.all' command)
    for (const cmd of Object.keys(Modernizer.allRules)) {
        context.subscriptions.push(vscode.commands.registerCommand(cmd, () => {
            processDocument(cmd as OCMCommand);
        }));
    }
    
    // Compound rules
    const ruleGroups: OCMCompoundRuleCommand[] = [
        'objective-clean.rules.all',
        'objective-clean.rules.enabled',
        'objective-clean.rules.literals.all',
        'objective-clean.rules.subscripts.all',
    ];
    for (const cmd of ruleGroups) {
        context.subscriptions.push(vscode.commands.registerCommand(cmd, () => {
            processDocument(cmd);
        }));
    }
}

// this method is called when your extension is deactivated
export function deactivate() {
}


webpack.config.js
//@ts-check

'use strict';

const path = require('path');

/**@type {import('webpack').Configuration}*/
const config = {
    target: 'node', // vscode extensions run in a Node.js-context 📖 -> https://webpack.js.org/configuration/node/
    mode: 'none', // this leaves the source code as close as possible to the original (when packaging we set this to 'production')

    entry: './src/extension.ts', // the entry point of this extension, 📖 -> https://webpack.js.org/configuration/entry-context/
    output: {
        // the bundle is stored in the 'dist' folder (check package.json), 📖 -> https://webpack.js.org/configuration/output/
        path: path.resolve(__dirname, 'dist'),
        filename: 'extension.js',
        libraryTarget: 'commonjs2'
    },
    devtool: 'nosources-source-map',
    externals: {
        vscode: 'commonjs vscode' // the vscode-module is created on-the-fly and must be excluded. Add other modules that cannot be webpack'ed, 📖 -> https://webpack.js.org/configuration/externals/
        // modules added here also need to be added in the .vsceignore file
    },
    resolve: {
        // support reading TypeScript and JavaScript files, 📖 -> https://github.com/TypeStrong/ts-loader
        extensions: ['.ts', '.js']
    },
    module: {
        rules: [
            {
                test: /\.ts$/,
                exclude: /node_modules/,
                use: [
                    {
                        loader: 'ts-loader'
                    }
                ]
            }
        ]
    }
};
module.exports = config;


src/modernizer.ts
//
//  modernizer.ts
//  Objective-Clean
//  
//  Created by Tanner Bennett on 2021-05-19
//  Copyright © 2021 Tanner Bennett. All rights reserved.
//

import * as vscode from 'vscode';
import { Scanner } from './scanner';

export type Rule = (input: string) => string;

export type OCMRuleCommand = 
    'objective-clean.rules.literals.nsnumber' |
    'objective-clean.rules.literals.nsdictionary' |
    'objective-clean.rules.literals.nsarray' |
    'objective-clean.rules.subscripts.indexed' |
    'objective-clean.rules.subscripts.keyed' |
    'objective-clean.rules.constants.cgrectzero' |
    'objective-clean.rules.init.allocinit-new' |
    'objective-clean.rules.init.initwithcapacity-new' |
    'objective-clean.rules.init.initwithframe0-new' |
    'objective-clean.rules.moderncomments';

export type OCMCompoundRuleCommand =
    'objective-clean.rules.all' |
    'objective-clean.rules.enabled' |
    'objective-clean.rules.literals.all' |
    'objective-clean.rules.subscripts.all';

export type OCMCommand = OCMRuleCommand | OCMCompoundRuleCommand;

class Modernizer implements vscode.TextDocumentContentProvider {
    static shared = new Modernizer();
    
    static allRules: { [cmd in OCMRuleCommand]: Rule } = {
        'objective-clean.rules.literals.nsnumber': Modernizer.NSNumberLiterals,
        'objective-clean.rules.literals.nsdictionary': Modernizer.NSDictionaryLiterals,
        'objective-clean.rules.literals.nsarray': Modernizer.NSArrayLiterals,
        'objective-clean.rules.subscripts.indexed': Modernizer.NSArrayAccessors,
        'objective-clean.rules.subscripts.keyed': Modernizer.NSDictionaryAccessors,
        'objective-clean.rules.constants.cgrectzero': Modernizer.CGRectZero,
        'objective-clean.rules.init.allocinit-new': Modernizer.AllocInitToNew,
        'objective-clean.rules.init.initwithcapacity-new': Modernizer.InitWithCapacityToNew,
        'objective-clean.rules.init.initwithframe0-new': Modernizer.InitWithFrame0ToNew,
        'objective-clean.rules.moderncomments': Modernizer.ModernComments,
    };
    
    static compoundRulesByCmd: { [cmd in OCMCompoundRuleCommand]: Rule[] } = {
        'objective-clean.rules.all': Object.values(Modernizer.allRules),
        'objective-clean.rules.literals.all': [
            Modernizer.NSNumberLiterals,
            Modernizer.NSArrayLiterals,
            Modernizer.NSDictionaryLiterals,
        ],
        'objective-clean.rules.subscripts.all': [
            Modernizer.NSArrayAccessors,
            Modernizer.NSDictionaryAccessors,
        ],
        'objective-clean.rules.enabled': [] // Special-cased in rulesForCmd
    };
    
    static get enabledRulesCommands(): Rule[] {
        const ruleCommands = Object.entries(Modernizer.allRules);
        // Filter all rules by whether they're enabled
        const rules = ruleCommands.filter(([cmd, rule]) => {
            // Split command by . and remove last component
            let components = cmd.split('.');
            const id = components.pop()!;
            const key = components.join('.');
            return vscode.workspace.getConfiguration(key).get(id);
        });
        
        return rules.map(([cmd, rule]) => rule);
    }
    
    static rulesForCmd(cmd: OCMCommand): Rule[] {
        const compoundCmd = cmd as OCMCompoundRuleCommand;
        const singleCmd = cmd as OCMRuleCommand;
        
        if (compoundCmd == 'objective-clean.rules.enabled') {
            return Modernizer.enabledRulesCommands;
        }
        
        return Modernizer.compoundRulesByCmd[compoundCmd] ?? [Modernizer.allRules[singleCmd]];
    }
    
    static NSDictionaryObjsKeysListToPairs(input: string): string {
        const scanner = new Scanner(input);
        
        // Gather keys and values as pairs to be swapped
        const replacements: [string,string][] = [];
        while (!scanner.isAtEnd) {
            const obj = scanner.scanUpToStringAfterTags(',', true);
            scanner.scanString(',');
            const key = scanner.scanUpToStringAfterTags(',', true); // Always ends in ", nil]"
            scanner.scanString(',');
            
            if (obj && key) {
                replacements.push([obj, key]);
            } else {
                // Couldn't parse input; abort
                return input;
            }
        }
        
        for (const pair of replacements) {
            const obj = pair[0], key = pair[1];
            // Replace the key first since it is after the object.
            // Segment is 'obj, key,' here with pair ['obj', 'key']
            input = input.replace(key, obj);
            // Segment is now 'obj, obj,'
            input = input.replace(`${obj},`, `${key}:`);
            // Segment is now 'key: obj,'
        }
        
        return input;
        
        // 1: The spacing
        // 2: The object
        // 3: The key
        // const pattern = /(\s+)(.+), *(.+),/;
        // let match = input.match(pattern);
        // if (!match) return input;
        
        // // Find each `  A, B,` and convert it to `  B: A,`
        // while (match) {
        //     const spacing = match[1];
        //     const obj = match[2];
        //     const key = match[3];
        //     input = input.replace(match[0], `${spacing}${key}: ${obj},`);
        //     match = input.match(pattern);
        // }
        
        // return input;
    }

    static NSDictionaryLiterals(text: string): string {
        // 1: Everything before the objs+keys
        // 2: 'Mutable' | ''
        // 3: The objects and keys
        // 4: Everything after the objs+keys
        const pattern = /(\[NS(Mutable)?Dictionary\s+dictionaryWithObjectsAndKeys:)(.+?)\s*(nil\s*\])/s;
        let matches = text.match(pattern);
        
        while (matches) {
            const prefix = matches[1];
            const mutability = matches[2] !== undefined ? '.mutableCopy' : '';
            const objsAndKeys = matches[3];
            const ending = matches[4];
            
            // Abort if we couldn't match objs & keys
            if (!objsAndKeys) return text;
            
            const keyValuePairs = Modernizer.NSDictionaryObjsKeysListToPairs(objsAndKeys);
            text = text
                .replace(prefix, '@{')
                .replace(objsAndKeys, keyValuePairs)
                .replace(ending, `}${mutability}`);
            
            matches = text.match(pattern);
        };
        
        return text;
    }

    static NSArrayLiterals(text: string): string {
        // Handle arrayWithObject:
        text = text.replace(/\[NSArray\s+arrayWithObject:\s*(.+)\s*\]/, '@[$1]');
        
        // 1: Everything before the values
        // 2: 'Mutable' | ''
        // 3: The objects and keys
        // 4: Everything after the objs+keys
        const pattern = /(\[NS(Mutable)?Array\s+arrayWithObjects: *)(.+)(\s*nil\s*\])/s;
        let matches = text.match(pattern);
        
        while (matches) {
            
            const prefix = matches[1];
            const mutability = matches[2] !== undefined ? '.mutableCopy' : '';
            const values = matches[3];
            const ending = matches[4];
            
            // Abort if we couldn't match values
            if (!values) return text;
            
            text = text
                .replace(prefix, '@[')
                .replace(ending, `]${mutability}`);
                
            matches = text.match(pattern);
        }
        
        return text;
    }

    static NSNumberLiterals(text: string): string {
        return text
        // Number literals like @5 or @3.14
        .replace(/\[NSNumber numberWith\w+:\s*([\d\.]+)\]/g, '@$1')
        // Boolean literals like @YES
        .replace(/\[NSNumber numberWith\w+:\s*(YES|NO)\]/g, '@$1')
        // Wrapper literals like @(count)
        .replace(/\[NSNumber numberWith\w+:\s*([\w\.]+)\]/g, '@($1)');
    }

    static NSDictionaryAccessors(text: string): string {
        // Dictionary accessors
        return text
            .replace(/\[([^\[]+)\s+objectForKey:([^\]]+)\]/g, '$1[$2]')
            .replace(/\[([^\[]+)\s+setObject:(.+) forKey:([^\]]+)\]/g, '$1[$3] = $2');
    }

    static NSArrayAccessors(text: string): string {
        // Array accessors
        return text
            .replace(/\[([^\[]+)\s+objectAtIndex:([^\]]+)\]/g, '$1[$2]')
            .replace(/\[([^\[]+)\s+replaceObjectAtIndex:(.+) withObject:([^\]]+)\]/g, '$1[$2] = $3');
    }

    static CGRectZero(text: string): string {
        return text.replace(/CGRectMake\((?:0+.?0*f?\s*,\s*){3}0+.?0*f?\s*\)/g, 'CGRectZero');
    }
    
    static AllocInitToNew(text: string): string {
        return text.replace(/\[\[([\w_][\w\d_$]+) alloc\] init\]/g, '[$1 new]');
    }
    
    static InitWithCapacityToNew(text: string): string {
        return text.replace(/\[\[([\w_][\w\d_$]+) alloc\] initWithCapacity:.+\]/g, '[$1 new]');
    }
    
    static InitWithFrame0ToNew(text: string): string {
        return text.replace(/\[\[([\w_][\w\d_$]+) alloc\] initWithFrame:\s*(?:CGRectZero|CGRectMake\((?:0+.?0*f?\s*,\s*){3}0+.?0*f?\s*\))\s*]/g, '[$1 new]');
    }
    
    static ModernComments(input: string): string {
        let scanner = new Scanner(input);
        
        while (scanner.scanUpToPattern(/ *\/\*/)) {
            const start = scanner.location;
            const pre = scanner.scanned;
            const openingWithWhitespace = scanner.scanPattern(/ *\/\*\*?/) as string;
            const leadingWhitespace = openingWithWhitespace.split('/*')[0];
            const content = scanner.scanUpToString('*/');
            scanner.scanString('*/');
            const post = scanner.remaining;
            
            if (!content) {
                // Couldn't parse input; abort
                return input;
            }
            
            // Split content into individual lines
            let lines = content.split('\n').map(l => l.trimStart());
            // Remove empty lines from start and end
            if (!lines[0].length) {
                lines.shift();
            }
            if (!lines[lines.length-1].length) {
                lines.pop();
            }
            // Remove leading * and don't add leading space as most
            // comemnts like this will have a space after the *,
            // or just add a space
            lines = lines.map(l => l.startsWith('*') ? l.substr(1) : (' ' + l));
            
            // Choose slash style based on whether /** /* was used
            const slashes = openingWithWhitespace.endsWith('**') ? '///' : '//';
            const newContent = lines.map(l => `${leadingWhitespace}${slashes}${l}`).join('\n');
            input = pre + newContent + post;
            scanner = new Scanner(input);
            // Offset scanning by the length of the content, -1 for each line
            scanner.location = start + newContent.length - lines.length;
        }
        
        return input;
    }
    
    /** Applies the given rules to the text */
    static applyRules(text: string, rules: Rule[]) {
        for (const rule of rules) {
            text = rule(text);
        }
        
        return text;
    }

    /** Applies all rules to given text */
    static modernize(text: string): string {
        for (const rule of Object.values(Modernizer.allRules)) {
            text = rule(text);
        }
        
        return text;
    }
    
    static applyCommandToText(cmd: OCMCommand, text: string): string {
        return Modernizer.applyRules(text, Modernizer.rulesForCmd(cmd));
    }
    
    static applyCommandToTextEditor(cmd: OCMCommand, editor: vscode.TextEditor) {
        const fileContents = editor.document.getText();
        const modifiedContents = Modernizer.applyCommandToText(cmd, fileContents);
        Modernizer.shared.applyChangesToTextEditor(modifiedContents, editor);
    }
    
    provideTextDocumentContent(diff: vscode.Uri): vscode.ProviderResult<string> {
        // Grab command and original file URI
        const cmd = diff.query.split('cmd=')[1] as OCMCommand;
        const orig = vscode.Uri.file(diff.path);
        
        // Get text from original document, modify it, and return it
        return vscode.workspace.openTextDocument(orig).then(async (document) => {
            const fileContents = document.getText();
            const modifiedContents = Modernizer.applyCommandToText(cmd, fileContents);
            
            // Present information dialog asking to confirm changes
            const msg = "Apply changes? This will overwrite the open editor.";
            vscode.window.showInformationMessage(msg, "Yes", "No").then((choice) => {
                if (choice == "Yes") {
                    this.applyChangesToFile(modifiedContents, orig, diff);
                }
            });
            
            return modifiedContents;
        });
    }
    
    areURIsEqual(first?: vscode.Uri, second?: vscode.Uri): boolean {
        if (!first || !second) { return false; }
        return first.toString(true) == second.toString(true);
    }
    
    applyChangesToFile(changes: string, uri: vscode.Uri, diff: vscode.Uri): boolean {
        // Check the active editor first
        if (this.areURIsEqual(uri, vscode.window.activeTextEditor?.document.uri)) {
            this.applyChangesToTextEditor(changes, vscode.window.activeTextEditor!, diff);
            return true;
        }
        
        // Check all other editors
        for (const editor of vscode.window.visibleTextEditors) {
            if (this.areURIsEqual(uri, editor.document.uri)) {
                this.applyChangesToTextEditor(changes, editor, diff);
                return true;
            }
        }
        
        return false;
    }
    
    applyChangesToTextEditor(changes: string, editor: vscode.TextEditor, diff?: vscode.Uri) {
        const document = editor.document;
        editor.edit(async editBuilder => {
            // Replace entire file contents
            var firstLine = document.lineAt(0);
            var lastLine = document.lineAt(document.lineCount - 1);
            var fullTextRange = new vscode.Range(firstLine.range.start, lastLine.range.end);
            editBuilder.replace(fullTextRange, changes);
            
            // Close diff view, if any, by bringing it to the front then closing the active editor
            if (diff) {
                await vscode.window.showTextDocument(diff, { preview: true, preserveFocus: false });
                return vscode.commands.executeCommand('workbench.action.closeActiveEditor');
            }
        });
    }
}

export default Modernizer;
